In [63]:
import pandas as pd
import numpy as np
import re

# General Info

In [64]:
shark = pd.read_csv('/Users/sarahpierre-louis/Downloads/GSAF5.csv',encoding = "ISO-8859-1")
shark
df= pd.DataFrame(shark)
df

Case Number                                               Date  Year  \
0     2016.09.18.c                                          18-Sep-16  2016   
1     2016.09.18.b                                          18-Sep-16  2016   
2     2016.09.18.a                                          18-Sep-16  2016   
3       2016.09.17                                          17-Sep-16  2016   
4       2016.09.15                                          16-Sep-16  2016   
5     2016.09.15.R                                          15-Sep-16  2016   
6       2016.09.11                                          11-Sep-16  2016   
7       2016.09.07                                          07-Sep-16  2016   
8       2016.09.06                                          06-Sep-16  2016   
9     2016.09.05.b                                          05-Sep-16  2016   
10    2016.09.05.a                                          05-Sep-16  2016   
11      2016.09.04                                          04-Sep-16  2016   
12      2016.09.01                                          01-Sep-16  2016   
13    2016.08.29.b                                          29-Aug-16  2016   
14    2016.08.29.a                                          29-Aug-16  2016   
15      2016.08.27                                          27-Aug-16  2016   
16      2016.08.25                                          25-Aug-16  2016   
17      2016.08.07                                          07-Aug-16  2016   
18      2016.08.06                                          06-Aug-16  2016   
19      2016.08.04                                          04-Aug-16  2016   
20      2016.07.29                                          29-Jul-16  2016   
21    2016.07.28.R                                          28-Jul-16  2016   
22      2016.07.28                                          28-Jul-16  2016   
23      2016.07.27                                          27-Jul-16  2016   
24      2016.07.26                                          26-Jul-16  2016   
25      2016.07.24                                          24-Jul-16  2016   
26    2016.07.23.b                                          23-Jul-16  2016   
27    2016.07.23.a                                          23-Jul-16  2016   
28      2016.07.20                                          20-Jul-16  2016   
29      2016.07.17                                          17-Jul-16  2016   
...            ...                                                ...   ...   
5962       ND.0031                                       World War II     0   
5963       ND.0030                                        Before 1905     0   
5964       ND.0028                            A few years before 1938     0   
5965       ND.0027                                            No date     0   
5966       ND.0026                                        Early 1930s     0   
5967       ND.0025                                        Before 1927     0   
5968       ND.0024                                Between 1918 & 1939     0   
5969       ND.0023                                            No date     0   
5970       ND.0022                                            No date     0   
5971       ND.0021                                            No date     0   
5972       ND.0020                                         1920 -1923     0   
5973       ND.0019                                        Before 1921     0   
5974       ND.0018                                        Before 1911     0   
5975       ND.0017                                        Before 1921     0   
5976       ND.0016                                        Before 1921     0   
5977       ND.0015                                        Before 1917     0   
5978       ND.0014                        Before          17-Jul-1916     0   
5979       ND.0013  No date (3 days after preceding incident) & pr...     0   
5980       ND.0012                        Before         

In [65]:
df.shape

(5992, 24)

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 24 columns):
Case Number               5992 non-null object
Date                      5992 non-null object
Year                      5992 non-null int64
Type                      5992 non-null object
Country                   5949 non-null object
Area                      5590 non-null object
Location                  5496 non-null object
Activity                  5465 non-null object
Name                      5792 non-null object
Sex                       5425 non-null object
Age                       3311 non-null object
Injury                    5965 non-null object
Fatal (Y/N)               5973 non-null object
Time                      2779 non-null object
Species                   3058 non-null object
Investigator or Source    5977 non-null object
pdf                       5992 non-null object
href formula              5991 non-null object
href                      5989 non-null object
C

In [67]:
df1=df.copy()

In [68]:
null_col = df.isnull().sum()
null_col

Case Number                  0
Date                         0
Year                         0
Type                         0
Country                     43
Area                       402
Location                   496
Activity                   527
Name                       200
Sex                        567
Age                       2681
Injury                      27
Fatal (Y/N)                 19
Time                      3213
Species                   2934
Investigator or Source      15
pdf                          0
href formula                 1
href                         3
Case Number.1                0
Case Number.2                0
original order               0
Unnamed: 22               5991
Unnamed: 23               5990
dtype: int64

In [69]:
#calculate the percentage of empty rows

null_col = df.isnull().sum()/df.shape[0]*100 #df.shape[0] : to take only the first value of shape
null_col = round(null_col)
null_col


Case Number                 0.0
Date                        0.0
Year                        0.0
Type                        0.0
Country                     1.0
Area                        7.0
Location                    8.0
Activity                    9.0
Name                        3.0
Sex                         9.0
Age                        45.0
Injury                      0.0
Fatal (Y/N)                 0.0
Time                       54.0
Species                    49.0
Investigator or Source      0.0
pdf                         0.0
href formula                0.0
href                        0.0
Case Number.1               0.0
Case Number.2               0.0
original order              0.0
Unnamed: 22               100.0
Unnamed: 23               100.0
dtype: float64

In [70]:
null_col[null_col>70]

Unnamed: 22    100.0
Unnamed: 23    100.0
dtype: float64

In [71]:
#drop the column with more than 70%
drop_cols = null_col[null_col>70].index
df1.drop(drop_cols, axis = 1, inplace= True) #drop doesnt drop but print without the drop


In [72]:
df1.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order'],
      dtype='object')

# Case Number

In [73]:
#look at the columns Case Number
df1[['Case Number.2', 'Case Number.1', 'Case Number']].T.duplicated()

Case Number.2    False
Case Number.1    False
Case Number      False
dtype: bool

In [74]:
df1[['Case Number.2', 'Case Number.1', 'Case Number']].sample(int(df1.shape[0]*0.9)).T.duplicated()
#df1.shape is a tuple, I select the nb of rows, I multiply it by 0.9 to get 99% of rows. I will get a float so I int() it then I sample it to get random rows

Case Number.2    False
Case Number.1    False
Case Number      False
dtype: bool

In [75]:
df1[['Case Number.2', 'Case Number.1', 'Case Number']].sample(int(df1.shape[0]*0.6)).T.duplicated()
#more than 60% but less than 99% of Case Number is in Case Number 1 and Case Number 2 

Case Number.2    False
Case Number.1    False
Case Number       True
dtype: bool

In [76]:
df1[['Case Number.2', 'Case Number.1', 'Case Number']].sample(10)

Case Number.2 Case Number.1   Case Number
913    2009.04.17.   2009.04.17.   2009.04.17.
2772  1982.00.00.b  1982.00.00.b  1982.00.00.b
4439    1943.10.26    1943.10.26    1943.10.26
88      2016.02.19    2016.02.19    2016.02.19
5118    1913.09.21    1913.09.21    1913.09.21
4211  1951.08.16.R  1951.08.16.R  1951.08.16.R
298     2014.08.06    2014.08.06    2014.08.06
1158  2007.05.07.b  2007.05.07.b  2007.05.07.b
5298    1902.04.00    1902.04.00    1902.04.00
5504  1886.08.00.c  1886.08.00.c  1886.08.00.c

# Year

In [77]:
df1.Year.describe()
#we have a lot of 0, they are much heavier than the max


count    5992.000000
mean     1925.204606
std       286.473712
min         0.000000
25%      1942.000000
50%      1975.000000
75%      2003.000000
max      2016.000000
Name: Year, dtype: float64

In [78]:
df1.Year.value_counts()

2015    139
2011    128
2014    125
0       124
2013    122
2008    121
2009    120
2012    117
2007    112
2006    103
2016    103
2005    103
2010    101
2000     97
1959     93
1960     93
2001     92
2004     92
2003     92
2002     88
1962     86
1961     78
1995     76
1964     66
1998     65
1999     65
1996     61
1963     61
1966     58
1997     57
       ... 
1785      1
1834      1
1791      1
1733      1
1721      1
1637      1
1617      1
77        1
5         1
1703      1
1755      1
1767      1
1771      1
1779      1
1787      1
1803      1
1749      1
1807      1
1811      1
1819      1
1805      1
1831      1
1555      1
1738      1
1859      1
1742      1
1758      1
1818      1
1822      1
1595      1
Name: Year, Length: 232, dtype: int64

In [79]:
df1.Year==0

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
        ...  
5962     True
5963     True
5964     True
5965     True
5966     True
5967     True
5968     True
5969     True
5970     True
5971     True
5972     True
5973     True
5974     True
5975     True
5976     True
5977     True
5978     True
5979     True
5980     True
5981     True
5982     True
5983     True
5984     True
5985     True
5986     True
5987     True
5988     True
5989     True
5990     True
5991     True
Name: Year, Length: 5992, dtype: bool

In [80]:
sum(df1.Year <1200)
old_rows_to_drop = df1.Year[df1.Year<1200].index
#the rows we need to drop
df2 = df1.drop(old_rows_to_drop)

In [81]:
df2.shape

(5865, 22)

# Date


In [82]:
df2.Date

0                      18-Sep-16
1                      18-Sep-16
2                      18-Sep-16
3                      17-Sep-16
4                      16-Sep-16
5                      15-Sep-16
6                      11-Sep-16
7                      07-Sep-16
8                      06-Sep-16
9                      05-Sep-16
10                     05-Sep-16
11                     04-Sep-16
12                     01-Sep-16
13                     29-Aug-16
14                     29-Aug-16
15                     27-Aug-16
16                     25-Aug-16
17                     07-Aug-16
18                     06-Aug-16
19                     04-Aug-16
20                     29-Jul-16
21                     28-Jul-16
22                     28-Jul-16
23                     27-Jul-16
24                     26-Jul-16
25                     24-Jul-16
26                     23-Jul-16
27                     23-Jul-16
28                     20-Jul-16
29                     17-Jul-16
          

In [83]:
df2[['Date', 'Year']].tail(20)

Date  Year
5845                      1755  1755
5846                      1748  1748
5847               17-Dec-1742  1742
5848      Reported 06-Apr-1738  1738
5849                      1733  1733
5850                 June 1721  1721
5851               26-Mar-1703  1703
5852                     1700s  1700
5853                     1700s  1700
5854                     1700s  1700
5855  Late 1600s Reported 1728  1642
5856             Reported 1638  1638
5857             Reported 1637  1637
5858             Reported 1617  1617
5859                      1642  1642
5860                      1595  1595
5861  Letter dated 10-Jan-1580  1580
5862                      1555  1555
5863                  Ca. 1554  1554
5864                  Ca. 1543  1543

In [84]:
import re

In [85]:
re.fullmatch(r'\d{4}', str(df2.Year[3804]))

<re.Match object; span=(0, 4), match='1960'>

In [86]:
#1. we drop the rows where we only have the year - eliminate the options that won't fit for sure
#2. create a new column : to get the month
def month_detection(value):
    value= str(value)
    if re.fullmatch(r'\d{4}', value):
        return -1 #means no month detected


In [87]:
month_detection(df2.Date[3804])

-1

In [88]:
df2.shape

(5865, 22)

In [89]:
#frequency analysis: to see what value is the most frequent one
df2.Date.tolist() #tolist to create a list from a series
[i.strip('Reportd ') for i in df2.Date.tolist()] #strip remove the letter onthe left and on the right side

['18-Sep-16',
 '18-Sep-16',
 '18-Sep-16',
 '17-Sep-16',
 '16-Sep-16',
 '15-Sep-16',
 '11-Sep-16',
 '07-Sep-16',
 '06-Sep-16',
 '05-Sep-16',
 '05-Sep-16',
 '04-Sep-16',
 '01-Sep-16',
 '29-Aug-16',
 '29-Aug-16',
 '27-Aug-16',
 '25-Aug-16',
 '07-Aug-16',
 '06-Aug-16',
 '04-Aug-16',
 '29-Jul-16',
 '28-Jul-16',
 '28-Jul-16',
 '27-Jul-16',
 '26-Jul-16',
 '24-Jul-16',
 '23-Jul-16',
 '23-Jul-16',
 '20-Jul-16',
 '17-Jul-16',
 '16-Jul-16',
 '16-Jul-16',
 '15-Jul-16',
 '14-Jul-2016',
 '08-Jul-2016',
 '08-Jul-16',
 '07-Jul-16',
 '07-Jul-16',
 '06-Jul-16',
 '04-Jul-16',
 '27-Jun-16',
 '25-Jun-16',
 '24-Jun-16',
 '23-Jun-16',
 '21-Jun-16',
 '21-Jun-16',
 '15-Jun-16',
 '15-Jun-16',
 '14-Jun-16',
 '11-Jun-16',
 '07-Jun-16',
 '05-Jun-16',
 '05-Jun-16',
 '04-Jun-16',
 '02-Jun-16',
 '02-Jun-16',
 '31-May-16',
 '29-May-16',
 '29-May-16',
 '22-May-16',
 '21-May-16',
 '21-May-16',
 '18-May-16',
 '15-May-16',
 '03-May-16',
 '02-May-16',
 '25-Apr-16',
 '23-Apr-16',
 '22-Apr-16',
 '19-Apr-16',
 '18-Apr-16',
 '

In [90]:
l = [k for i in df2.Date.tolist() for j in i.split() for k in j.split('-')]
l
# équivalent à 
# l= []
#for i in df2.Date.tolist():
#    i_split=i.split()
#    for j in i_split():
#       j_split= j.split('-')
#    for k in j_split:
#       l.append(k)

['18',
 'Sep',
 '16',
 '18',
 'Sep',
 '16',
 '18',
 'Sep',
 '16',
 '17',
 'Sep',
 '16',
 '16',
 'Sep',
 '16',
 '15',
 'Sep',
 '16',
 '11',
 'Sep',
 '16',
 '07',
 'Sep',
 '16',
 '06',
 'Sep',
 '16',
 '05',
 'Sep',
 '16',
 '05',
 'Sep',
 '16',
 '04',
 'Sep',
 '16',
 '01',
 'Sep',
 '16',
 '29',
 'Aug',
 '16',
 '29',
 'Aug',
 '16',
 '27',
 'Aug',
 '16',
 '25',
 'Aug',
 '16',
 '07',
 'Aug',
 '16',
 '06',
 'Aug',
 '16',
 '04',
 'Aug',
 '16',
 '29',
 'Jul',
 '16',
 '28',
 'Jul',
 '16',
 '28',
 'Jul',
 '16',
 '27',
 'Jul',
 '16',
 '26',
 'Jul',
 '16',
 '24',
 'Jul',
 '16',
 '23',
 'Jul',
 '16',
 '23',
 'Jul',
 '16',
 '20',
 'Jul',
 '16',
 '17',
 'Jul',
 '16',
 '16',
 'Jul',
 '16',
 '16',
 'Jul',
 '16',
 '15',
 'Jul',
 '16',
 'Reported',
 '14',
 'Jul',
 '2016',
 'Reported',
 '08',
 'Jul',
 '2016',
 '08',
 'Jul',
 '16',
 '07',
 'Jul',
 '16',
 '07',
 'Jul',
 '16',
 '06',
 'Jul',
 '16',
 '04',
 'Jul',
 '16',
 '27',
 'Jun',
 '16',
 '25',
 'Jun',
 '16',
 '24',
 'Jun',
 '16',
 '23',
 'Jun',
 '16',
 '

In [91]:
len(set(l)) #nb of different values that are in Date

442

REGEX

Step 1:
    
looking for [A-Za-z] {3,}

Step 2:
Remove those with these patterns 
["A-Za-z\.\?)&]* to remove "Anniversary", "Bitten", (?)...
\d{3,4} to remove 144 et 1595
\d{2,4}s to remove 1600s
\d{4}/\d{4} to remove 1843/1844
\d{4}\.\d{2}\. to remove 1844.07.16.R
\d{4}\.[a-z] 
 

In [92]:
set(l)

{'',
 '"Anniversary',
 '"Bitten',
 '"on',
 '&',
 '(?)',
 '(probably',
 '(same',
 '(soon',
 '.',
 '.Reported',
 '00',
 '01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '144',
 '15',
 '1543',
 '1554',
 '1555',
 '1580',
 '1595',
 '16',
 '1600s',
 '1617',
 '1637',
 '1638',
 '1642',
 '17',
 '1700s',
 '1703',
 '1721',
 '1728',
 '1733',
 '1738',
 '1742',
 '1748',
 '1749',
 '1755',
 '1758',
 '1764',
 '1767',
 '1771',
 '1776',
 '1779',
 '1785',
 '1787',
 '1788',
 '1791',
 '18',
 '1800',
 '1803',
 '1804',
 '1805',
 '1807',
 '1811',
 '1812',
 '1816',
 '1817',
 '1818',
 '1819',
 '1820s',
 '1822',
 '1825',
 '1826',
 '1827',
 '1828',
 '1829',
 '1830',
 '1831',
 '1832',
 '1834',
 '1835',
 '1836.00.',
 '1836.07.26.R',
 '1837',
 '1839',
 '1839/1840',
 '1840',
 '1841',
 '1842',
 '1843/1844)',
 '1844',
 '1844.07.16.R',
 '1845',
 '1846',
 '1847',
 '1848',
 '1849',
 '1850',
 '1851',
 '1852',
 '1853',
 '1854',
 '1855',
 '1856',
 '1858',
 '1859',
 '1860',
 '18

In [93]:
def month_detection(value):
    
    value= str(value)
    if re.fullmatch(r'[A-Za-z]{3,}', value):
        return value
    elif re.fullmatch(r'\d{4}', value):
        return -1
    elif re.fullmatch(r'[\"A-Za-z\.\(\)\?&]*', value):
        return -1
    elif re.fullmatch(r'\d{3,5}', value):
        return -1
    elif re.fullmatch(r'd{2,4}s', value):
        return -1
    elif re.fullmatch(r'\d{4}/\d{4}\)?', value):
        return -1
    elif re.fullmatch (r'\d{4}\.\d{2}\.', value):
        return -1
    elif re.fullmatch(r'\[A-Za-z]+ \d+', value):
        return -1
    else:
        return value

In [94]:
df2['Month']=df2.Date.apply(month_detection)
df2.Month

0                      18-Sep-16
1                      18-Sep-16
2                      18-Sep-16
3                      17-Sep-16
4                      16-Sep-16
5                      15-Sep-16
6                      11-Sep-16
7                      07-Sep-16
8                      06-Sep-16
9                      05-Sep-16
10                     05-Sep-16
11                     04-Sep-16
12                     01-Sep-16
13                     29-Aug-16
14                     29-Aug-16
15                     27-Aug-16
16                     25-Aug-16
17                     07-Aug-16
18                     06-Aug-16
19                     04-Aug-16
20                     29-Jul-16
21                     28-Jul-16
22                     28-Jul-16
23                     27-Jul-16
24                     26-Jul-16
25                     24-Jul-16
26                     23-Jul-16
27                     23-Jul-16
28                     20-Jul-16
29                     17-Jul-16
          

In [95]:
#it can be full name of the name (months) or abbreviation of the month (months_short)
#we create 2 groups, if it parts of 1 of the groups, we 

def month_detection(value):
    
    months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
    months_short = [i[:3] for i in months] #list with the first 3 letters of months
    value= str(value).strip()
    
    if re.fullmatch(r'[A-Za-z]{3,}', value):
        gr1 = re.fullmatch(r'[A-Za-z]{3,}', value).group(1)
        if (gr1 in months) | (gr1 in months_short):
            return gr1
        else:
            return -1
    elif re.fullmatch(r'\d{4}', value):
        return -1
    elif re.fullmatch(r'[\"A-Za-z\.\(\)\?&]*', value):
        return -1
    elif re.fullmatch(r'\d{3,5}', value):
        return -1
    elif re.fullmatch(r'd{2,4}s', value):
        return -1
    elif re.fullmatch(r'\d{4}/\d{4}\)?', value):
        return -1
    elif re.fullmatch (r'\d{4}\.\d{2}\.', value):
        return -1
    elif re.fullmatch(r'([A-Za-z]+) \d+', value):
        gr1 = re.fullmatch(r'([A-Za-z]+) \d+', value).group(1) #we put group(1) because we just want the month
        if (gr1 in months) | (gr1 in months_short):
            return gr1
        else:
            return -1
    elif re.fullmatch(r'\d{1,2}-(\w{2,3})-\d{2,4}', value):
            return re.fullmatch(r'\d{1,2}-(\w{2,3})-\d{2,4}', value).group(1)
    elif re.fullmatch(r'Reported \d{1,2}-(\w{2,3})-\d{2,4}', value):
            return re.fullmatch(r'Reported \d{1,2}-(\w{2,3})-\d{2,4}', value).group(1) 
    elif re.fullmatch(r'([A-Za-z]{3})-\d{2,4}', value):
            return re.fullmatch(r'([A-Za-z]{3})-\d{2,4}', value).group(1) 
    else:
        return value

In [96]:
month_detection('01-Feb-1943')

'Feb'

\d{1,2}-(\w{2,3})-\d{2,4} to get 08-Dec-1994 ou 8-12-94 et toutes les autres possibilités

In [97]:
df2['Month']=df2.Date.apply(month_detection)
df2.Month.value_counts()

Jul                                     614
Aug                                     566
Sep                                     515
Jan                                     491
Jun                                     470
Dec                                     416
Oct                                     407
Apr                                     400
Mar                                     391
Nov                                     388
May                                     364
Feb                                     353
-1                                      315
1970s                                     5
1700s                                     3
May-Jun-1965                              3
1980s                                     2
Ca. 1950                                  2
Ca. 1880                                  2
Ap                                        2
Mid-Aug-1886                              2
Ca. 1960                                  2
1954 (same day as  1954.00.00.f)

In [98]:
df3=df2.drop('Date', axis=1)

In [99]:
df3.shape

(5865, 22)

# Activity

In [100]:
df3.Activity.sample(50)

2481    The MV Dona Marilyn sank in Typhoon Unsang wit...
2208                                              Surfing
2799                                              Fishing
1284       Competing in the Woodvale Atlantic Rowing Race
4849                   Launching rowboat through the surf
5356                                    Fishing (Seining)
3524                                             Swimming
845                                                Diving
870                                              Swimming
3437                                             Swimming
2811                                              Surfing
4437                                  Swimming with board
490                                               Surfing
5735                                             Swimming
2613                            Hookah diving for abalone
5455                                 Fishing for bluefish
4475    The 6015-ton British ship Empire Avocet was to...
2870          

In [39]:
df3.Activity.value_counts()

Surfing                                                              903
Swimming                                                             804
Fishing                                                              407
Spearfishing                                                         321
Bathing                                                              150
Wading                                                               142
Diving                                                               111
Standing                                                              95
Snorkeling                                                            76
Scuba diving                                                          73
Body boarding                                                         54
Body surfing                                                          48
Swimming                                                              47
Treading water                                     

In [105]:
type(df3.Activity)

pandas.core.series.Series

In [106]:
Activity = list(df3.Activity)
type(Activity)

list

In [107]:
'in'.strip('ing')

''

In [198]:
l = [j.replace('ing','') for i in Activity for j in str(i).lower().split('&,./\\')]

bag_of_words= []

for i in l:
    if i not in bag_of_words:
        bag_of_words.append(i)
print(bag_of_words)

term_freq= []
blacklist = ['and', 'as', 'an', 'a', 'the', 'in', 'it', 'for', 'the', 'on']

for i in bag_of_words:
    if (i in l) and (i not in blacklist):
        term_freq.append(l.count(i))
print(term_freq)

dictionary = dict(zip(bag_of_words, term_freq))
print(dictionary)


['surf', 'fish', 'wad', 'swimm', 'kite surf', 'boogie board', 'body board', 'spearfish', 'snorkel', 'sup foil board', 'kayak', 'lobster', 'fish in alabama deep fish rodeo', 'nan', 'scuba div', 'fish for squid', 'fish for sharks', 'float', 'float in tube', 'stand', 'div', 'teas a shark', 'surf-ski', 'walk', 'sup', 'body surf', 'div for beche-de-mer', 'feed strays?', 'windsurf', 'kayak fish', 'paddle board', 'sea disaster', 'lobster fish', 'swimm / wad', 'fish for shrimp', 'photograph fish', 'stand-up paddleboard', 'scallop div on hookah', 'tread water', 'body board ', 'play in the water', 'body surf?', 'play', 'attempt to rescue a shark', 'surf ski ', 'photograph the shark', 'stand / snorkel', 'hand feed sharks', 'sitt in the water', 'transatlantic row', 'bodysurf', 'row', 'shark fish', 'surf & film dolphins', 'film', 'swimm after fall overboard', 'div / film', 'fish for blue sharks', 'canoe', 'stand in inner tube', 'body surf or boogie board', 'kayak ', 'feed fish', 'fish ', 'play with

In [143]:
dic_sorted = sorted(dictionary, key=dictionary.get, reverse=True)
for i in dic_sorted:
    print (i, dictionary[i])

surf 903
swimm 804
nan 506
fish 407
spearfish 321
bath 150
wad 142
div 111
stand 95
scuba div 80
snorkel 76
body board 57
body surf 48
swimm  47
boogie board 44
tread water 32
kayak 28
pearl div 28
free div 24
fell overboard 21
windsurf 18
shark fish 16
walk 15
sea disaster 14
float 12
surf-ski 12
kayak fish 12
row 12
canoe 12
fish  12
surf ski 12
boat 12
fish for sharks 11
surf fish 11
freediv 10
kite surf 9
sail 9
sitt on surfboard 9
freedom swimm 9
paddle board 8
div for abalone 8
div for trochus 8
play 7
fell into the water 7
wade fish 7
surf (sitt on his board) 7
free div for abalone 7
spearfish on scuba 7
sponge div 7
surf ski  6
spearfish / free div 6
murder 6
float on his back 6
skindiv 6
fish for mackerel 6
paddleski 5
fish on a boat 5
shipwreck 5
hard hat div 5
dangl feet in the water 5
stand-up paddleboard 4
kite board 4
free div / spearfish 4
spearfish (free div) 4
air disaster 4
jump 4
jumped into the water 4
lifesav drill 4
body-board 4
splash 4
shell div 4
paddl on surfb

washed overboard into school of fish 1
swimm with motor tube 1
in deep water about 100 yards from his ship 1
surf, but tread water 1
dragg banana seeds through the shallows 1
swimm ashore from fish boat swamped and sunk by a squall 1
paddl & sail from buenos aires to miami 1
surf on air mattress 1
clean hull of ship  1
slapped shark on tail as it swam by 1
"climb up to ship after repair the stern in water" 1
boat stopped to repair electric pump 1
spearfish, had fish on his spear 1
swimm, tow the shark 1
paddl in knee-deep water 1
swimm with goggles 1
fish, hold fish in his left hand 1
spearfish / swimm on surface 1
placed hand in disemboweled shark’s jaws 1
fish, hold fish 1
fish (troll) 1
fish, walk in river to cast 1
air/sea disaster pan-american airlines stratocruiser with 44 people onboard crashed into the sea 1
attempt to net shark in shark channel 1
m.v. tropical sank. sole survivor rode oil drums for 8 days without food or water. 1
compet in spearfish championship & tow dead fis

In [199]:
import pandas as p

In [200]:
df=pd.Series(l)
df.value_counts()

surf                                                                                                                                                                                                                                            903
swimm                                                                                                                                                                                                                                           804
nan                                                                                                                                                                                                                                             506
fish                                                                                                                                                                                                                                            407
spearfish               

In [201]:
#other option
df = pd.DataFrame(l)
df[0].value_counts()

surf                                                                                                                                                                                                                                            903
swimm                                                                                                                                                                                                                                           804
nan                                                                                                                                                                                                                                             506
fish                                                                                                                                                                                                                                            407
spearfish               

In [187]:
activity_nona = df3.Activity[df3.Activity.isna()== False]

In [188]:
l=[j.strip('.,/\\()&').lower() for i in activity_nona for j in i.split()]

In [202]:
words_df = pd.Series(l) #series and not dataframe because we dont have a column name
words_df.value_counts() #amount of words we have in our dataframe

surf                                                                                                                                                                                                                                            903
swimm                                                                                                                                                                                                                                           804
nan                                                                                                                                                                                                                                             506
fish                                                                                                                                                                                                                                            407
spearfish               

In [203]:
from nltk.corpus import stopwords   

stop_words = set(stopwords.words('english'))   

l = [w for w in l if not w in stop_words] 

words_df=pd.Series(l)

words_df.value_counts()

surf                                                                                                                                                                                                                                            903
swimm                                                                                                                                                                                                                                           804
nan                                                                                                                                                                                                                                             506
fish                                                                                                                                                                                                                                            407
spearfish               

In [204]:
bow = list(words_df.value_counts()[:10].index)

In [207]:
bow

['surf',
 'swimm',
 'nan',
 'fish',
 'spearfish',
 'bath',
 'wad',
 'div',
 'stand',
 'scuba div']

In [208]:
bow.remove('spearfishing')

ValueError: list.remove(x): x not in list

In [209]:
import numpy as np

lst = []
for i in df3.Activity: 
    lst1= [] #create an empty list
    for ind,j in enumerate (bow): #loop among the bag of words,  #enumerate gives a tuple with the index (ind is going to be 1, 2, 3) and j is swim, surf ...
        if j in str(i).lower(): #if swim is in swimming 
            lst1.append(j) #then append swim to lst1
            break #then I don't go further, if the first word was swim , I take swim and I stop
        if ind == len(bow)-1:
            lst1.append(-1) #if I dont have a value, the value will be -1, which is "other"
   
    lst.append(lst1[0]) #we append the first and only one element
df3['new_activity'] = lst
df3[['new_activity', 'Activity']]


new_activity                                           Activity
0            surf                                            Surfing
1            surf                                            Surfing
2            surf                                            Surfing
3            surf                                            Surfing
4            surf                                            Surfing
5            fish                                            Fishing
6             wad                                             Wading
7           swimm                                           Swimming
8            surf                                       Kite surfing
9              -1                                    Boogie boarding
10           surf                                            Surfing
11             -1                                      Body boarding
12           fish                                       Spearfishing
13           surf                                            Surfing
14           surf                                            Surfing
15           surf                                            Surfing
16            wad                                             Wading
17             -1                                         Snorkeling
18             -1                                  SUP Foil boarding
19           surf                                            Surfing
20          swimm                                           Swimming
21          swimm                                           Swimming
22             -1                                           Kayaking
23             -1                                         Lobstering
24           surf                                            Surfing
25           surf                                            Surfing
26           surf                                            Surfing
27           fish                                       Spearfishing
28           fish                                            Fishing
29           fish              Fishing in Alabama Deep Fishing Rodeo
...           ...                                                ...
5835        swimm                                           Swimming
5836          nan                                                NaN
5837         surf                                            Surfing
5838           -1                                             Murder
5839          nan                                                NaN
5840         fish                                            Fishing
5841         bath                                            Bathing
5842        swimm                                           Swimming
5843           -1  Fell overboard from a frigate & was swallowed ...
5844        swimm                                           Swimming
5845          nan                                                NaN
5846          div                                       Pearl diving
5847        swimm                                           Swimming
5848        swimm                                           Swimming
5849          nan                                                NaN
5850        swimm                                           Swimming
5851        swimm                                           Swimming
5852          nan                                                NaN
5853         bath                                            Bathing
5854         bath                                            Bathing
5855           -1                                     Went overboard
5856          nan                                                NaN
5857          wad                                             Wading
5858          nan                                                NaN
5859        swimm                                          Swimming 
5860           -1  Ship lay at anchor & man was working on its ru...
586

In [196]:
# another way to do it (less pretty)

lst = []
for i in df3.Activity:
    lst.append([j for j in bow if j in str(i).lower()])
df3['new_activity'] = lst
df3[['new_activity', 'Activity']]


new_activity                                           Activity
0      [surfing, ]                                            Surfing
1      [surfing, ]                                            Surfing
2      [surfing, ]                                            Surfing
3      [surfing, ]                                            Surfing
4      [surfing, ]                                            Surfing
5      [fishing, ]                                            Fishing
6       [, wading]                                             Wading
7     [swimming, ]                                           Swimming
8      [surfing, ]                                       Kite surfing
9               []                                    Boogie boarding
10     [surfing, ]                                            Surfing
11              []                                      Body boarding
12     [fishing, ]                                       Spearfishing
13     [surfing, ]                                            Surfing
14     [surfing, ]                                            Surfing
15     [surfing, ]                                            Surfing
16      [, wading]                                             Wading
17              []                                         Snorkeling
18              []                                  SUP Foil boarding
19     [surfing, ]                                            Surfing
20    [swimming, ]                                           Swimming
21    [swimming, ]                                           Swimming
22              []                                           Kayaking
23              []                                         Lobstering
24     [surfing, ]                                            Surfing
25     [surfing, ]                                            Surfing
26     [surfing, ]                                            Surfing
27     [fishing, ]                                       Spearfishing
28     [fishing, ]                                            Fishing
29     [fishing, ]              Fishing in Alabama Deep Fishing Rodeo
...            ...                                                ...
5835  [swimming, ]                                           Swimming
5836            []                                                NaN
5837   [surfing, ]                                            Surfing
5838            []                                             Murder
5839            []                                                NaN
5840   [fishing, ]                                            Fishing
5841   [, bathing]                                            Bathing
5842  [swimming, ]                                           Swimming
5843     [, shark]  Fell overboard from a frigate & was swallowed ...
5844  [swimming, ]                                           Swimming
5845            []                                                NaN
5846    [diving, ]                                       Pearl diving
5847  [swimming, ]                                           Swimming
5848  [swimming, ]                                           Swimming
5849            []                                                NaN
5850  [swimming, ]                                           Swimming
5851  [swimming, ]                                           Swimming
5852            []                                                NaN
5853   [, bathing]                                            Bathing
5854   [, bathing]                                            Bathing
5855            []                                     Went overboard
5856            []                                                NaN
5857    [, wading]                                             Wading
5858            []                                                NaN
5859  [swimming, ]                                          Swimming 
5860            

In [ ]:
remove Name

In [ ]:
Regex select only the numbers cannot be more than 120

In [ ]:
for injury did the same that we did for activity

In [ ]:
can remove the time

In [ ]:
investigator and le reste can drop them

In [21]:
before = len(shark)
shark = shark.drop_duplicates()
after = len(shark)
print('Number of duplicate records dropped: ', str(before - after))

Number of duplicate records dropped:  0


In [22]:
drop_cols = ['Unnamed: 22', 'Unnamed: 23', 'Case Number.1', 'Case Number.2', 'href formula']
shark1 = shark.drop(drop_cols, axis=1)

In [23]:
shark1

Case Number                                               Date  Year  \
0     2016.09.18.c                                          18-Sep-16  2016   
1     2016.09.18.b                                          18-Sep-16  2016   
2     2016.09.18.a                                          18-Sep-16  2016   
3       2016.09.17                                          17-Sep-16  2016   
4       2016.09.15                                          16-Sep-16  2016   
5     2016.09.15.R                                          15-Sep-16  2016   
6       2016.09.11                                          11-Sep-16  2016   
7       2016.09.07                                          07-Sep-16  2016   
8       2016.09.06                                          06-Sep-16  2016   
9     2016.09.05.b                                          05-Sep-16  2016   
10    2016.09.05.a                                          05-Sep-16  2016   
11      2016.09.04                                          04-Sep-16  2016   
12      2016.09.01                                          01-Sep-16  2016   
13    2016.08.29.b                                          29-Aug-16  2016   
14    2016.08.29.a                                          29-Aug-16  2016   
15      2016.08.27                                          27-Aug-16  2016   
16      2016.08.25                                          25-Aug-16  2016   
17      2016.08.07                                          07-Aug-16  2016   
18      2016.08.06                                          06-Aug-16  2016   
19      2016.08.04                                          04-Aug-16  2016   
20      2016.07.29                                          29-Jul-16  2016   
21    2016.07.28.R                                          28-Jul-16  2016   
22      2016.07.28                                          28-Jul-16  2016   
23      2016.07.27                                          27-Jul-16  2016   
24      2016.07.26                                          26-Jul-16  2016   
25      2016.07.24                                          24-Jul-16  2016   
26    2016.07.23.b                                          23-Jul-16  2016   
27    2016.07.23.a                                          23-Jul-16  2016   
28      2016.07.20                                          20-Jul-16  2016   
29      2016.07.17                                          17-Jul-16  2016   
...            ...                                                ...   ...   
5962       ND.0031                                       World War II     0   
5963       ND.0030                                        Before 1905     0   
5964       ND.0028                            A few years before 1938     0   
5965       ND.0027                                            No date     0   
5966       ND.0026                                        Early 1930s     0   
5967       ND.0025                                        Before 1927     0   
5968       ND.0024                                Between 1918 & 1939     0   
5969       ND.0023                                            No date     0   
5970       ND.0022                                            No date     0   
5971       ND.0021                                            No date     0   
5972       ND.0020                                         1920 -1923     0   
5973       ND.0019                                        Before 1921     0   
5974       ND.0018                                        Before 1911     0   
5975       ND.0017                                        Before 1921     0   
5976       ND.0016                                        Before 1921     0   
5977       ND.0015                                        Before 1917     0   
5978       ND.0014                        Before          17-Jul-1916     0   
5979       ND.0013  No date (3 days after preceding incident) & pr...     0   
5980       ND.0012                        Before         


# Fatal Y/N

In [9]:
shark1['Fatal (Y/N)'].value_counts()

N          4315
Y          1552
UNKNOWN      94
 N            8
N             1
F             1
n             1
#VALUE!       1
Name: Fatal (Y/N), dtype: int64

In [10]:
shark1['Fatal (Y/N)'] = shark1['Fatal (Y/N)'].str.replace(' N', 'N').replace('F', 'N').replace('n', 'N')

In [11]:
shark1['Fatal (Y/N)'].value_counts()

N          4325
Y          1552
UNKNOWN      94
N             1
#VALUE!       1
Name: Fatal (Y/N), dtype: int64

# Injury

In [142]:
shark['Injury'].value_counts()

FATAL                                                                                                                   729
Survived                                                                                                                 97
Foot bitten                                                                                                              82
No injury                                                                                                                77
Leg bitten                                                                                                               71
Left foot bitten                                                                                                         48
No details                                                                                                               43
Right foot bitten                                                                                                        39
Hand bit

In [30]:
for i in shark1['Injury']:
    if ("No injury" or "No injurys" or "No inury,") in i:
        shark1['Injury'] = shark1['Injury'].str.replace(i, "No injury")
    elif ("Fatal" or "FATAL," or "Fatal.")in i:
        shark1['Injury'] = shark1['Injury'].str.replace(i, "FATAL")
    elif ("bitten" or "bitten,") in i:
        shark1['Injury'] = shark1['Injury'].str.replace(i, "bitten")

        print(shark1['Injury'])

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

Name: Injury, Length: 5992, dtype: object
0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

Name: Injury, Length: 5992, dtype: object
0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                                  Minor injury to ankle
17            

Name: Injury, Length: 5992, dtype: object
0                                   Minor injury to thigh
1                                    Lacerations to hands
2                                Lacerations to lower leg
3                            Struck by fin on chest & leg
4                                               No injury
5                                               No injury
6                                     Minor injury to arm
7                Severe lacerations to shoulder & forearm
8                                                   FATAL
9              Lacerations & punctures to lower right leg
10                No inury, board broken in half by shark
11                                                 bitten
12                            Two toes broken & lacerated
13                              Lacerations to right foot
14                                  Minor injury to ankle
15                                                 bitten
16                            

TypeError: argument of type 'float' is not iterable

In [31]:
shark1['Injury'].value_counts()

FATAL                                                                                                                            729
bitten                                                                                                                           658
No injury                                                                                                                        290
Survived                                                                                                                          97
No details                                                                                                                        43
FATAL, body not recovered                                                                                                         24
Minor injury                                                                                                                      22
Foot lacerated                                                       

# Date

In [18]:
shark1['Date'].value_counts()

1957                    11
1942                     9
1956                     8
1941                     7
1950                     7
1958                     7
1949                     6
No date                  6
1955                     5
1970s                    5
1954                     5
05-Oct-03                5
12-Apr-01                5
28-Jul-95                5
Oct-60                   5
1940                     5
No date, Before 1963     5
1959                     5
Aug-56                   5
20-Sep-15                4
1876                     4
1945                     4
1898                     4
1938                     4
09-Jul-94                4
1960                     4
1961                     4
28-Dec-14                4
1952                     4
23-Jan-70                4
                        ..
02-Apr-13                1
12-Sep-00                1
21-Sep-65                1
May-92                   1
24-Sep-10                1
24-Jan-83                1
R

In [33]:
pip install python-dateutil

Note: you may need to restart the kernel to use updated packages.


In [37]:
from dateutil import parser
>>> parser.parse("12-Apr-01")
shark1['Date'].shark1['Date'](2001, 4, 12, 0, 0)
>>> parser.parse("1957")
shark1['Date'].shark1['Date'](1957, 0, 0, 0, 0)
>>> parser.parse("May-92")
shark1['Date'].shark1['Date'](1992, 5, 0, 0, 0)


>>> parser.parse("12-Apr-01").strftime('%Y-%m-%d')
'2001-04-12'
>>> parser.parse("1957").strftime('%Y-%m-%d')
'1957-01-01'
>>> parser.parse("May-92").strftime('%Y-%m-%d')
'1992-05-01'




AttributeError: 'Series' object has no attribute 'shark1'

In [119]:
shark3= str(shark1['Date'])
pattern = '\W+'
shark3 = re.sub(pattern,'', shark3)
shark3

'018Sep16118Sep16218Sep16317Sep16416Sep16515Sep16611Sep16707Sep16806Sep16905Sep161005Sep161104Sep161201Sep161329Aug161429Aug161527Aug161625Aug161707Aug161806Aug161904Aug162029Jul162128Jul162228Jul162327Jul162426Jul162524Jul162623Jul162723Jul162820Jul162917Jul165962WorldWarII5963Before19055964Afewyearsbefore19385965Nodate5966Early1930s5967Before19275968Between191819395969Nodate5970Nodate5971Nodate5972192019235973Before19215974Before19115975Before19215976Before19215977Before19175978Before17Jul19165979Nodate3daysafterprecedingincidentpr5980Before19Jul19135981Before19115982Circa18625983Before19065984Before19065985Before19065986Before19065987Before19035988Before1903598919001905599018831889599118451853NameDateLength5992dtypeobject'

In [39]:
shark1['Date'] = shark1['Date'].values.astype('datetime64[M]')

ValueError: Error parsing datetime string "18-Sep-16" at position 3

# Name

In [26]:
shark1['Name'].value_counts()

male                                                                                495
female                                                                               85
boy                                                                                  23
2 males                                                                              14
boat                                                                                 14
Anonymous                                                                            10
child                                                                                10
sailor                                                                               10
a sailor                                                                              7
girl                                                                                  7
males                                                                                 6
fisherman                       

In [11]:
shark1.loc[(shark1['Name']=='male') | (shark1['Name']=='boy'), 'Sex'] = 'M'
shark1.loc[(shark1['Name']=='female')|(shark1['Name']=='girl') ,'Sex'] = 'F'

# Sex

In [12]:
shark1['Sex'].value_counts()

M      4841
F       589
N         1
lli       1
M         1
.         1
Name: Sex, dtype: int64

In [9]:
shark1 = shark1.rename(columns={'Sex ':'Sex'})

# Type

In [37]:
shark1['Type'].value_counts()

Unprovoked    4606
Provoked       557
Invalid        519
Boat           310
Name: Type, dtype: int64

In [33]:
shark1['Type'] = shark1['Type'].str.replace('Boating',('Boat'))

In [36]:
shark1['Type'] = shark1['Type'].str.replace('Sea Disaster',('Unprovoked'))

# Activity

In [43]:
shark1['Activity'].value_counts()

Surfing                                                                                                    904
Swimming                                                                                                   819
Fishing                                                                                                    414
Spearfishing                                                                                               321
Bathing                                                                                                    153
Wading                                                                                                     144
Diving                                                                                                     123
Standing                                                                                                    97
Snorkeling                                                                                                  77
S

In [12]:
for i in shark1['Activity']:
    if "surfing" in i:
        shark1['Activity'] = shark1['Activity'].str.replace(i, "Surfing")
    elif "Fishing" in i:
        shark1['Activity']= shark1['Activity'].str.replace(i, "Fishing")
    
        print(shark1['Activity'])

0                                                 Surfing
1                                                 Surfing
2                                                 Surfing
3                                                 Surfing
4                                                 Surfing
5                                                 Fishing
6                                                  Wading
7                                                Swimming
8                                                 Surfing
9                                         Boogie boarding
10                                                Surfing
11                                          Body boarding
12                                           Spearfishing
13                                                Surfing
14                                                Surfing
15                                                Surfing
16                                                 Wading
17            

TypeError: argument of type 'float' is not iterable

In [13]:
shark1['Activity'].value_counts()

Surfing                                                                                                      907
Swimming                                                                                                     819
Fishing                                                                                                      415
Spearfishing                                                                                                 321
Bathing                                                                                                      153
Wading                                                                                                       144
Diving                                                                                                       123
Standing                                                                                                      97
Snorkeling                                                                                      

# Age


In [41]:
shark1['Age'].value_counts()

17                148
18                145
19                138
20                136
15                135
16                134
21                115
22                113
24                103
25                101
14                 97
13                 91
26                 80
23                 80
27                 78
28                 77
29                 75
30                 74
12                 70
35                 64
32                 64
10                 51
31                 50
40                 50
38                 47
34                 45
43                 43
36                 41
33                 40
39                 37
                 ... 
7      &    31      1
X                   1
34 & 19             1
30 or 36            1
M                   1
25 to 35            1
33 & 26             1
32 & 30             1
Elderly             1
9 months            1
81                  1
>50                 1
A.M.                1
21 & ?              1
72        

In [43]:
shark1[['Age']] = shark1[['Age']].fillna(0)

In [140]:
import re 

In [141]:
s= str(shark1['Age'])
pattern = '[A-Za-z]*'
shark1 = re.sub(pattern,'', s)
print(shark1)

TypeError: string indices must be integers

In [3]:
import pandas as pd

In [14]:
l = ['find a string', 'without for regex', 'in python']

df=pd.DataFrame(l)

In [15]:
'regex' in l

False

In [16]:
['regex' in i for i in l ]

[False, True, False]

In [17]:
df[0] [df[0].str.contains('regex')]= 'regex'
df

0
0  find a string
1          regex
2      in python

In [9]:
df[0].str.contains('for')

0    False
1      NaN
2    False
Name: 0, dtype: object